pandas/core/indexes/multi.py

```diff
@@ -1774,14 +1784,20 @@ def to_frame(self, index: bool = True, name=lib.no_default) -> DataFrame:
         else:
             idx_names = self.names
 
+        idx_names = [
+            level if name is None else name for level, name in enumerate(idx_names)
+        ]
+
+        if not allow_duplicates and len(set(idx_names)) != len(idx_names):
+            raise ValueError(
+                "Cannot create duplicate column labels if allow_duplicates is False"
+            )
+
         # Guarantee resulting column order - PY36+ dict maintains insertion order
         result = DataFrame(
-            {
-                (level if lvlname is None else lvlname): self._get_level_values(level)
-                for lvlname, level in zip(idx_names, range(len(self.levels)))
-            },
+            {level: self._get_level_values(level) for level in range(len(self.levels))},
             copy=False,
-        )
+        ).set_axis(idx_names, axis=1)
```

Comment:
this requires another operation, just set it in place directly

Reply:
Can't be done in the constructor, because the constructor will not accept columns argument with dictionary data. And I have to use dictionary data, otherwise I lose the column dtypes.

I was under the mistaken impression that Pandas was trying to move away from in-place opterations, hence the set_axis. I have changed it to in-place.

Reply:
not in internal code where it makes sense (but yes in user facing code), just *never* the `inplace` keyword. setting columns is a standard operation.